### PR TITLE
added type mapping and support for different types

### PR DESCRIPTION
**Pull request by spuetz**
As there has been no reading and writing conversion methods for the PointCloud2 buffer, I implemented these methods. Also the new convertPointCloud2ToPointCloud supports different data types.

### CODE REVIEW

File: sensor_msgs/include/sensor_msgs/point_cloud_conversion.h [hunk 1]
```diff
@@ -116,13 +117,14 @@ static inline bool convertPointCloudToPointCloud2 (const sensor_msgs::PointCloud
 static inline bool convertPointCloud2ToPointCloud (const sensor_msgs::PointCloud2 &input, sensor_msgs::PointCloud &output)
 {
   // Get all fields size and check if we have non-float32 values
-  for (size_t d = 0; d < input.fields.size (); ++d)
+ /* for (size_t d = 0; d < input.fields.size (); ++d)
```
Comment (tfoote) [inline sensor_msgs/include/sensor_msgs/point_cloud_conversion.h]: Don't leave unused code.

File: sensor_msgs/include/sensor_msgs/point_cloud_conversion.h [hunk 2]
```diff
@@ -155,19 +160,22 @@ static inline bool convertPointCloud2ToPointCloud (const sensor_msgs::PointCloud
...
     {
       if ((int)input.fields[d].offset == x_offset || (int)input.fields[d].offset == y_offset || (int)input.fields[d].offset == z_offset)
         continue;
-      memcpy (&output.channels[cur_c++].values[cp], &input.data[cp * input.point_step + input.fields[d].offset], sizeof (float));
+      output.channels[cur_c++].values[cp] = sensor_msgs::convert<float>(&input.data[cp * input.point_step + input.fields[d].offset], input.fields[d].datatype);
     }
   }
   return (true);
 }
+
```
Comment (tfoote) [inline sensor_msgs/include/sensor_msgs/point_cloud_conversion.h]: Remove white space changes only.

File: sensor_msgs/include/sensor_msgs/point_field_conversion.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  template<> struct asType<sensor_msgs::PointField::UINT8>   { typedef uint8_t  type; };
+  template<> struct asType<sensor_msgs::PointField::INT16>   { typedef int16_t  type; };
+  template<> struct asType<sensor_msgs::PointField::UINT16>  { typedef uint16_t type; };
+  template<> struct asType<sensor_msgs::PointField::INT32>   { typedef int32_t  type; };
+  template<> struct asType<sensor_msgs::PointField::UINT32>  { typedef uint32_t type; };
+  template<> struct asType<sensor_msgs::PointField::FLOAT32> { typedef float    type; };
+  template<> struct asType<sensor_msgs::PointField::FLOAT64> { typedef double   type; };
+  
+  /*!
+   * \Type to enum mapping.
```
Comment (tfoote) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h:70]: Is this necessary?

**Comment by tfoote** [comment]
Overall this looks good. Could you add a basic unit test which will test a few different type conversions. 

Also have you done any performance comparisons? In particular does this change the float performance. 

I had a few inline comments too. 

@vrabaud could you take a look too?

File: sensor_msgs/include/sensor_msgs/point_field_conversion.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
+/*
+ * Robot Operating System code by the University of Osnabrück
+ * Copyright (C) 2015 University of Osnabrück
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
```
Comment (vrabaud) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: Does this have to be GPL ?

File: sensor_msgs/include/sensor_msgs/point_field_conversion.h [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+   *  specified by the given datatype parameter to the given tampleate type and returns it.
+   * \param data_ptr    pointer into the point cloud 2 buffer
+   * \param datatype    datatype specified as in the mapping above
+   * \tparam return type
+   */
+  template<typename T>
+    T convert(const unsigned char* data_ptr, const unsigned char datatype){
+      switch(datatype){
+        case sensor_msgs::PointField::INT8:
+          return (T)*((asType<sensor_msgs::PointField::INT8>::type const *)data_ptr);
```
Comment (vrabaud) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: use a static_cast

File: sensor_msgs/include/sensor_msgs/point_field_conversion.h [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+   * \param data_ptr    pointer into the point cloud 2 buffer
+   * \param datatype    datatype specified as in the mapping above
+   * \param value       the value to insert
+   * \tparam type of the value to insert
+   */
+  template<typename T>
+    void insert(const unsigned char* data_ptr, const unsigned char datatype, T value){
+      switch(datatype){
+        case sensor_msgs::PointField::INT8:
+          *((asType<sensor_msgs::PointField::INT8>::type*)data_ptr) = (asType<sensor_msgs::PointField::INT8>::type)value;
```
Comment (vrabaud) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: use static_cast

File: sensor_msgs/include/sensor_msgs/point_field_conversion.h [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  * \brief  This file provides a type to enum mapping for the different
+  *         PointField types and methods to read and write in 
+  *         a PointCloud2 buffer for the different PointField types.
+  * \author Sebastian Pütz
+  */
+namespace sensor_msgs{
+  /*!
+   * \Enum to type mapping.
+   */
+  template<int> struct asType {};
```
Comment (vrabaud) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: I'd rather see a more specific naming like pointFieldTypeAsType

File: sensor_msgs/include/sensor_msgs/point_field_conversion.h [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  
+  /*!
+   * \Converts a value at the given pointer position interpreted as the datatype
+   *  specified by the given datatype parameter to the given tampleate type and returns it.
+   * \param data_ptr    pointer into the point cloud 2 buffer
+   * \param datatype    datatype specified as in the mapping above
+   * \tparam return type
+   */
+  template<typename T>
+    T convert(const unsigned char* data_ptr, const unsigned char datatype){
```
Comment (vrabaud) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: too generic: how about pointFieldConvert ?

File: sensor_msgs/include/sensor_msgs/point_field_conversion.h [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. All advertising materials mentioning features or use of this software
```
Comment (vrabaud) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: that seems too strong. @tfoote , any advice ? This is not the BSD license. @spuetz , you were asked to add that clause ?

**Comment by vrabaud** [comment]
except for the license, things have been modified properly (mostly, getting naming less generic) so I'm good. Thx. @tfoote , please advise on the non-BSD license

File: sensor_msgs/include/sensor_msgs/point_field_conversion.h (hunk 9 above)
Comment (gavanderhoorn) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: Isn't this 'just' the [4-clause](https://en.wikipedia.org/wiki/BSD_licenses#4-clause_license_.28original_.22BSD_License.22.29) BSD license? With the 'obnoxious BSD advertising clause'?

Comment (vrabaud) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: Right, it's obnoxious :) We usually choose the 3-clause one, cf http://wiki.ros.org/DevelopersGuide#Licensing

Comment (gavanderhoorn) [inline sensor_msgs/include/sensor_msgs/point_field_conversion.h]: I know :). I was just commenting on your "_This is not the BSD license_".

**Comment by spuetz** [comment]
The license is now the standard 3-clause BSD. Since a conversion of PointCloud2 to a PointCloud is just not good, and if you do a conversion, it should support every PointCloud2 also with other data types than float, the small performance difference should not be important. If one had a problem with the performance, he or she should use only PointCloud2 in the whole pipeline. I added these code, because i was wondering that there wasn't such a mapping of the enumeration types to the data types. I use this code in the https://github.com/spuetz/laser_assembler to do not first convert into a PointCloud to assemble clouds and than back. Now I directly assemble PointClouds2 and also support organized PointClouds2. I also think this code could be useful for the PointCloud2 iterator in jade-devel to handle the sizing automatically and everywhere where you build a PointCloud2 or read from it.

**Comment by tfoote** [comment]
This is looking good to me. i'll probably merge it against indigo as hydro is EOL. 

@vrabaud can you read it over once more?

File: sensor_msgs/include/sensor_msgs/point_cloud_conversion.h [hunk 10]
```diff
@@ -115,15 +116,6 @@ static inline bool convertPointCloudToPointCloud2 (const sensor_msgs::PointCloud
     */ 
 static inline bool convertPointCloud2ToPointCloud (const sensor_msgs::PointCloud2 &input, sensor_msgs::PointCloud &output)
 {
-  // Get all fields size and check if we have non-float32 values
-  for (size_t d = 0; d < input.fields.size (); ++d)
-  {
-    if (input.fields[d].datatype != sensor_msgs::PointField::FLOAT32)
-    {
-      std::cerr << boost::str(boost::format("sensor_msgs::PointCloud accepts only float32 values, but field %d (%s) has field type %d!")%(int)d% input.fields[d].name%input.fields[d].datatype) << std::endl;
```
Comment (vrabaud) [inline sensor_msgs/include/sensor_msgs/point_cloud_conversion.h:123]: Removing that condition would change the behavior: before, RGB as uchar could not be put into a PointCloud (by being converted to float32). Is that a behavior we want ? @tfoote , probably not

**Comment by vrabaud** [comment]
something obvious: you need to close that pull request and open it up for the indigo-devel branch, as hydro is end-of-lifed.

**Pull request closed without merging**